pkg/operator/controller/canary/controller.go

```diff
@@ -0,0 +1,177 @@
+	reconciler := &reconciler{
+		Config: config,
+		client: mgr.GetClient(),
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+
+	// Only trigger a reconcile request for the canary controller via events for the default ingress controller.
+	defaultIcPredicate := predicate.NewPredicateFuncs(func(meta metav1.Object, object runtime.Object) bool {
+		return meta.GetName() == manifests.DefaultIngressControllerName
+	})
+
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}, defaultIcPredicate); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Reconcile ensures that the canary controller's resources
+// are in the desired state.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	result := reconcile.Result{}
+	errors := []error{}
+
+	if err := r.ensureCanaryNamespace(); err != nil {
+		errors = append(errors, err)
+		return result, utilerrors.NewAggregate(errors)
+	}
+
+	haveDs, daemonset, err := r.ensureCanaryDaemonSet()
+	if err != nil {
+		errors = append(errors, err)
+		return result, utilerrors.NewAggregate(errors)
+	}
+	if !haveDs {
+		errors = append(errors, fmt.Errorf("failed to get canary daemonset"))
+		return result, utilerrors.NewAggregate(errors)
+	}
+
+	trueVar := true
+	daemonsetRef := metav1.OwnerReference{
+		APIVersion: "apps/v1",
+		Kind:       "daemonset",
+		Name:       daemonset.Name,
+		UID:        daemonset.UID,
+		Controller: &trueVar,
+	}
+
+	haveService, service, err := r.ensureCanaryService(daemonsetRef)
+	if err != nil {
+		errors = append(errors, err)
+		return result, utilerrors.NewAggregate(errors)
+	}
+	if !haveService {
+		errors = append(errors, fmt.Errorf("failed to get canary service"))
+		return result, utilerrors.NewAggregate(errors)
+	}
+
+	if haveRoute, _, err := r.ensureCanaryRoute(service); err != nil {
+		errors = append(errors, err)
+		return result, utilerrors.NewAggregate(errors)
+	} else if !haveRoute {
+		errors = append(errors, fmt.Errorf("failed to get canary route"))
+		return result, utilerrors.NewAggregate(errors)
+	}
+
+	return result, nil
+}
+
+// Config holds all the things necessary for the controller to run.
+type Config struct {
+	Namespace   string
+	CanaryImage string
+}
+
+// reconciler handles the actual canary reconciliation logic in response to
+// events.
+type reconciler struct {
+	Config
+
+	client client.Client
+}
+
+// Switch the current RoutePort that the route points to.
+// Use this function to periodically update the canary route endpoint
+// to verify if the router has wedged.
+func (r *reconciler) rotateRouteEndpoint(service *corev1.Service, current *routev1.Route) (*routev1.Route, error) {
+	updated, err := cycleServicePort(service, current)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to rotate route port: %v", err)
```

Comment:
What are our rules (or existing behaviour) for capitalising error message? (Do we even have any?)

Reply:
Not sure if it's formally documented anywhere, but I'll be sure to stick to lowercase.

Reply:
The most formal documentation I can find is https://github.com/golang/go/wiki/CodeReviewComments#error-strings.  Lower-case for errors and upper-case for logs is the prevalent convention in the Go world.